jax/dtypes.py

```diff
@@ -200,37 +198,39 @@ def _jax_type(value):
-  # promotion lattice. See testPromotionTableLattice for details of this.
+def _type_promotion_lattice():
   b1, u1, u2, u4, u8, i1, i2, i4, i8, bf, f2, f4, f8, c4, c8, i_, f_, c_ = _jax_types
-  #  b1, u1, u2, u4, u8, i1, i2, i4, i8, bf, f2, f4, f8, c4, c8, s*, f*, c*
-  return np.array([
-    [b1, u1, u2, u4, u8, i1, i2, i4, i8, bf, f2, f4, f8, c4, c8, i_, f_, c_],  # b1
-    [u1, u1, u2, u4, u8, i2, i2, i4, i8, bf, f2, f4, f8, c4, c8, u1, f_, c_],  # u1
-    [u2, u2, u2, u4, u8, i4, i4, i4, i8, bf, f2, f4, f8, c4, c8, u2, f_, c_],  # u2
-    [u4, u4, u4, u4, u8, i8, i8, i8, i8, bf, f2, f4, f8, c4, c8, u4, f_, c_],  # u4
-    [u8, u8, u8, u8, u8, f_, f_, f_, f_, bf, f2, f4, f8, c4, c8, u8, f_, c_],  # u8
-    [i1, i2, i4, i8, f_, i1, i2, i4, i8, bf, f2, f4, f8, c4, c8, i1, f_, c_],  # i1
-    [i2, i2, i4, i8, f_, i2, i2, i4, i8, bf, f2, f4, f8, c4, c8, i2, f_, c_],  # i2
-    [i4, i4, i4, i8, f_, i4, i4, i4, i8, bf, f2, f4, f8, c4, c8, i4, f_, c_],  # i4
-    [i8, i8, i8, i8, f_, i8, i8, i8, i8, bf, f2, f4, f8, c4, c8, i8, f_, c_],  # i8
-    [bf, bf, bf, bf, bf, bf, bf, bf, bf, bf, f4, f4, f8, c4, c8, bf, bf, c4],  # bf
-    [f2, f2, f2, f2, f2, f2, f2, f2, f2, f4, f2, f4, f8, c4, c8, f2, f2, c4],  # f2
-    [f4, f4, f4, f4, f4, f4, f4, f4, f4, f4, f4, f4, f8, c4, c8, f4, f4, c4],  # f4
-    [f8, f8, f8, f8, f8, f8, f8, f8, f8, f8, f8, f8, f8, c8, c8, f8, f8, c8],  # f8
-    [c4, c4, c4, c4, c4, c4, c4, c4, c4, c4, c4, c4, c8, c4, c8, c4, c4, c4],  # c4
-    [c8, c8, c8, c8, c8, c8, c8, c8, c8, c8, c8, c8, c8, c8, c8, c8, c8, c8],  # c8
-    [i_, u1, u2, u4, u8, i1, i2, i4, i8, bf, f2, f4, f8, c4, c8, i_, f_, c_],  # s*
-    [f_, f_, f_, f_, f_, f_, f_, f_, f_, bf, f2, f4, f8, c4, c8, f_, f_, c_],  # f*
-    [c_, c_, c_, c_, c_, c_, c_, c_, c_, c4, c4, c4, c8, c4, c8, c_, c_, c_],  # c*
-  ])
-
-_type_promotion_table = _make_type_promotion_table()
+  return {
+    b1: [i_],
+    u1: [i2, u2], u2: [i4, u4], u4: [i8, u8], u8: [f_],
+    i_: [u1, i1], i1: [i2], i2: [i4], i4: [i8], i8: [f_],
+    f_: [bf, f2, c_], bf: [f4], f2: [f4], f4: [f8, c4], f8: [c8],
+    c_: [c4], c4: [c8], c8: [],
+  }
+
+def _make_lattice_upper_bounds():
+  lattice = _type_promotion_lattice()
+  upper_bounds = {node: {node} for node in lattice}
+  for n in lattice:
+    while True:
+      new_upper_bounds = set().union(*(lattice[b] for b in upper_bounds[n]))
+      if n in new_upper_bounds:
+        raise ValueError(f"cycle detected in type promotion lattice for node {n}")
+      if not new_upper_bounds.difference(upper_bounds[n]):
+        break
+      upper_bounds[n] |= new_upper_bounds
+  return upper_bounds
+_lattice_upper_bounds = _make_lattice_upper_bounds()
+
+@functools.lru_cache(512)
+def _least_upper_bound(*nodes):
+  common_upper_bounds = set.intersection(*(_lattice_upper_bounds[n] for n in nodes))
+  least_upper_bounds = common_upper_bounds - set.union(*(_lattice_upper_bounds[n] - {n}
```

Comment:
It is not obvious to me when reading this that it computes LUB. If CUB(a, b) are the common upper bounds of a and b, then LUB(a, b) is defined as {c in CUB(a, b) | forall d in CUB(a, b). c <= d}. By definition this set is either empty or  a singleton (if <= is an order). 

It would be more satisfying if the code were more obviously correct. BTW, if we pre-compute the LUB before we start then we can afford to have a less efficient but more obviously correct implementation. 

Perhaps this is an academic question, since this code will not be used to compute LUB for arbitrary lattices, and I am sure it works for our lattice.

Reply:
I believe this code does exactly that. This is the reasoning:
- `c < d` if and only if `d ∈ UB(c) - {c}`
- Therefore, any node `d ∈ UB(c) - {c}` given `c ∈ CUB(a,b)` is greater than some other node in `CUB(a,b)`, and thus cannot be the least upper bound.
- It follows that removing all such nodes `d` from `CUB(a, b)` leaves only the least upper bound, so long as the input graph is actually a lattice.

If there is more than one remaining node, then the two are unorderable, and the input graph must not be a lattice.

Reply:
I added a comment describing this reasoning... let me know if it's clear.

Reply:
BTW, I think we can merge this. The discussion at this point is more fun than useful. 

But the proof is still tricky. LUB is by definition the unique element in CUB that is <= all CUB. In bullet 2 above you subtract from CUB nodes `d` that cannot be LUB. But I do not think that it follows that if you are left with a single node it must be LUB (i.e., bullet 3 is not obviously true). 

You would need to remove from CUB all nodes `d` such that there exists `c` in CUB with `not (d <= c)`. It is true that if `c < d` then `not (d <= c)` but not the other way around (unless you have a total order). 

The last statement "If there is more than one remaining node, then the two are unorderable, and the input graph must not be a lattice" also I do not think that is strictly true.

Reply:
Hmmm.... I'm out of practice in my proof formalisms :grin:. I'll think about this some more, but I think we can somehow show the third point so long as `<=` is transitive across partially ordered sets.

Reply:
OK, how about this:
- We want `LUB(a, b) = {c ∈ CUB(a, b) | forall d in CUB(a, b). c <= d}`
- `c <= d` implies `d ∈ UB(c)` (I think this is true, otherwise `c` and `d` are unorderable?)
- So we want `LUB(a, b) = {c ∈ CUB(a, b) | forall d in CUB(a, b). d ∈ UB(c)}`
- In other words `LUB(a, b) = {c ∈ CUB(a, b) | cardinality[CUB(a, b)] = cardinality[UB(c)]}`
So, all we need is to compute the cardinality of each node `c` in `CUB(a, b)` and choose the one whose upper bounds have the same cardinality as `CUB(a, b)`.

Reply:
Alright, I think I landed on an approach that's more obviously correct, and is about 20% faster to boot. See the explanation in the code.

Reply:
Your second bullet is true by definition of UB. 

Very nice, I believe this proof.